huntsman/observatory.py

```diff
@@ -232,202 +233,67 @@ def autofocus_cameras(self, *args, **kwargs):
-                    "Saturated with short exposure, waiting 30 seconds before next exposure")
-                max_num_exposures += 1
-                time.sleep(30)
-
-        # Add a bias exposure
-        for cam_name in camera_list:
-            exptimes[cam_name].append(0 * u.second)
-
-        # Record how many exposures we took
-        num_exposures = flat_obs.current_exp_num
-
-        # Take darks (just last ten)
-        for i in range(num_exposures):
-            self.logger.debug("Slewing to dark-field coords: {}".format(flat_obs.field))
-            self.mount.set_target_coordinates(flat_obs.field)
-            self.mount.slew_to_target()
-            self.status()
-
-            for cam_name in camera_list:
-
-                try:
-                    exptime = exptimes[cam_name][i]
-                except KeyError:
-                    break
-
-                camera = self.cameras[cam_name]
-
-                filename = "{}/flats/{}/{}/{}.{}".format(
-                    image_dir,
-                    camera.uid,
-                    flat_obs.seq_time,
-                    'dark_{:02d}'.format(flat_obs.current_exp_num),
-                    camera.file_extension)
-
-                # Take picture and wait for result
-                camera_event = camera.take_observation(
-                    flat_obs,
-                    fits_headers,
-                    filename=filename,
-                    exptime=exptime,
-                    dark=True
-                )
-
-                camera_events[cam_name] = {
-                    'event': camera_event,
-                    'filename': filename,
-                }
-
-            # Will block here until done exposing on all cameras
-            while not all([info['event'].is_set() for info in camera_events.values()]):
-                self.logger.debug('Waiting for dark-field image')
-                time.sleep(1)
-
-    def take_morning_flats(*args, **kwargs):
-        '''
-
-        '''
-        raise NotImplementedError('Morning flats not implemented yet.')
+        if camera_names is None:
+            cameras_all = self.cameras
+        else:
+            cameras_all = {c: self.cameras[c] for c in camera_names}
+
+        # Obtain the filter order
+        filter_order = self.config['filterwheel']['flat_field_order']
+        if self.past_midnight:  # If it's the morning, order is reversed
+            filter_order = filter_order[::-1]
+
+        exptimes_dark = {c: [0 * u.second] for c in cameras_all.keys()}  # Add a bias exposure
+        for filter_name in filter_order:
+
+            # Get a dict of cameras that have this filter
+            cameras = {}
+            for cam_name, cam in cameras_all.items():
+                if cam.filterwheel is None:
+                    if cam.filter_type == filter_name:
+                        cameras[cam_name] = cam
+                elif filter_name in cam.filterwheel.filter_names:
+                    cameras[cam_name] = cam
```

Comment:
As above, there needs to be an exact match here, which may cause problems. There is no convenient way to ask a `Filterwheel` "hey, do have a filter whose name starts with this?", though there should be. This is a ugly workaround for that deficiency:
```suggestion
                else: 
                    with suppress(ValueError):
                        if cam.filterwheel._parse_position(filter_name):
                            cameras[cam_name] = cam
```

You'd need a `from contextlib import suppress` too.

Reply:
That is quite ugly. I'm really not sure I like this idea of using the first letter of the filter names to identify them... sounds potentially error-prone for someone who is not aware. If we want to group filters for flat fielding then I suggest another mechanism is implemented.

Reply:
It's not actually matching the first letter, it looks for the whole string that it's given to be present at the start of the full filter name. It just happens that we've got filters that are of nominally the same type that only have the first letter in common (`g`, `r`). If we rename them things like `sloan_g_astrodon_gen2_03` and `sloan_g_chroma_08` (which we should perhaps do anyway) then we could select on `sloan_g`.

As for not being aware, well, it's fully documented in the `Filterwheel` class docs ;-) This is how the filterwheels normally work when you request a filter by name, & is consequently how the `filtername` attribute of `Observation` works.